README.md

```diff
@@ -20,76 +22,174 @@ This document defines an extension to the service types supported in the [DID Co
 
-The `id` property of the endpoint MUST be the [multibase][multibase] representation of the Blake256 output calculated from the Base64 encoding of the resource dereferenced by the URIs in `serviceEndpoint`.
-The multibase chosen must yield values that do not contain invalid characters as per the definition of the `id` property in the DID specification, i.e., the resulting `id` must still be a valid URI conforming to [RFC3986][rfc3986].
+### Object Canonicalization and Hashing
 
-Hence, calling `M` the multibase encoding operation of some data, `H` the Blake256 hashing, and `B` the binary representation of some information, the service `id` for a given object `O` is `M(H(B(O)))`.
+The `id` property of the endpoint MUST be the [multibase][multibase] representation of the Blake2b-256 output calculated from the Base64 encoding of the [canonical representation][rfc8785] of the resource dereferenced by the URIs in `serviceEndpoint`.
+The multibase chosen MUST yield values that do not contain invalid characters as per the definition of the `id` property in the DID specification, i.e., the resulting `id` MUST still be a valid URI conforming to [RFC3986][rfc3986].
 
+The retrieved resource MUST be canonicalized before being hashed, according to the [RFC 8785 specification][rfc8785].
+This canonicalization step is required to ensure that two semantically-equivalent services do not hash to two different values.
+
+Hence, calling `M` the multibase encoding operation of some data, `H` the Blake2b-256 hashing, `B` the binary representation of some information, and `N` the canonicalization step, the service `id` for a given object `O` is `M(H(B(N(O))))`.
 For example, with the object `O` being the example `serviceEndpoint` shown above, and the multibase `M` being `base64urlpad`, the resulting service endpoint looks like the following:
 
 ```json
 {
-  "id": "did:kilt:4pqDzaWi3w7TzYzGnQDyrasK6UnyNnW6JQvWRrq6r8HzNNGy#Uc1JU0UF9iDfjaRkgHCFG2Rc5jki-cuhlgbEQcjN6-g0=",
+  "id": "did:kilt:4pqDzaWi3w7TzYzGnQDyrasK6UnyNnW6JQvWRrq6r8HzNNGy#UHfpCR8mCNP5FvNRjN7rLHm7DA8fm7bqB6Pd4fGjaJ4Y=",
   "type": [
     "KiltTransferAssetRecipientV1"
   ],
   "serviceEndpoint": [
-    "https://ipfs.io/ipfs/QmfLCY7Jxa4rRhpQ1FK3srtyM6C5ZLk4YmXfS7nm4Ur2GG"
+    "https://ipfs.io/ipfs/QmYbmJiuUtuNjFA7L7HVCgCCK2GFbxZ2VttFgTvurrfZff"
   ]
 }
 ```
 
-## Security considerations
+A Typescript snippet showing how to derive the service ID for the example document above is shown below:
+
+```ts
+import { blake2AsU8a } from '@polkadot/util-crypto'
+import * as multibase from 'multibase'
+import canonicalize from 'canonicalize'
+
+const doc = `
+{
+  "polkadot:b0a8d493285c2df73290dfb7e61f870f/slip44:434": {
+    "EJDj2GKnx89HTzUkGW8Rk9RoYUmAJHPM8aacWFp3fi1gYUQ": {
+      "description": "Personal account"
+    }
+  },
+  "polkadot:411f057b9107718c9624d6aa4a3f23c1/slip44:2086": {
+    "4nvZhWv71x8reD9gq7BUGYQQVvTiThnLpTTanyru9XckaeWa": {
```

Comment:
Why it is required to change the format of the JSON breaking backward compatibility?

Reply:
Unfortunately it's because arrays are not sorted as part of the canonicalisation process: https://www.rfc-editor.org/rfc/rfc8785#section-3.2.3-5.2.

The JSON spec does not have a concept for unordered set, hence arrays are considered ordered data structures, and its elements not re-arranged. That's why the change, unfortunately.

Reply:
Got it. But do we really want to treat a list of accounts per chain as an unordered set? Because UI will usually display accounts as the list, a user may want to put more important accounts to the top.

Reply:
That's a conversation we might want to have. For instance URLs for a service endpoint are typically treated as an unordered set. In this case, we do have an optional `description` field which would let the user specify additional details. I would see this field as being present most of the time, and each address being dedicated to something different, potentially. I do not see the case why a user would expose two addresses for the same purpose, e.g., receiving simple transfers, where one address would be more important than the other. Could you see a case like that? Or have you seen such a case?

TL;DR: I think multiple addresses for the same asset class will usually be used for different purposes, which makes them not naturally "sortable" between them anyway. If two addresses are present for the very same purpose, I would like to know why a user might/could prefer one over the other.